packages/core/src/controller/Controller.ts

```diff
@@ -159,16 +169,158 @@ export default class Controller {
+
+    if (results !== undefined && meta?.errorPolicy === 'soft') return;
+
+    return meta?.error as any;
+  };
 
-  getError = <E extends EndpointInterface>(
+  getResponse = <E extends Pick<EndpointInterface, 'key' | 'schema'>>(
     endpoint: E,
-    ...args: readonly [...Parameters<E>, State<unknown>]
-  ): error
+    ...rest:
+      | readonly [...Parameters<E['key']>, State<unknown>]
+      | readonly [null, State<unknown>]
+  ): {
+    data: DenormalizeNullable<E['schema']>;
+    suspend: boolean;
+    found: boolean;
+    expiresAt: number;
+  } => {
+    const state = rest[rest.length - 1] as State<unknown>;
+    const args = rest.slice(0, rest.length - 1) as Parameters<E['key']>;
+    const activeArgs = args?.[0] !== null;
+    const key = activeArgs ? endpoint.key(...args) : '';
+    const cacheResults = activeArgs && state.results[key];
+    const schema = endpoint.schema;
 
-  */
+    const results = this.getResults(
+      endpoint.schema,
+      cacheResults,
+      args,
+      state.indexes,
+    );
+
+    if (!endpoint.schema || !schemaHasEntity(endpoint.schema)) {
+      return {
+        data: results,
+        suspend: false,
+        found: cacheResults !== undefined,
+        expiresAt: selectMeta(state, key)?.expiresAt || 0,
+      } as {
+        data: DenormalizeNullable<E['schema']>;
+        suspend: boolean;
+        found: boolean;
+        expiresAt: number;
+      };
+    }
+    // Warn users with bad configurations
+    /* istanbul ignore next */
+    if (process.env.NODE_ENV !== 'production' && schema && isEntity(schema)) {
+      if (Array.isArray(results)) {
+        throw new Error(
+          `fetch key ${key} has list results when single result is expected`,
+        );
+      }
+      if (typeof results === 'object') {
+        throw new Error(
+          `fetch key ${key} has object results when single result is expected`,
+        );
+      }
+    }
+
+    if (activeArgs && !this.globalCache.results[key])
+      this.globalCache.results[key] = new WeakListMap();
+
+    // second argument is false if any entities are missing
+    // eslint-disable-next-line prefer-const
+    const [data, found, suspend, resolvedEntities] = denormalize(
+      results,
+      schema,
+      state.entities,
+      this.globalCache.entities,
+      activeArgs ? this.globalCache.results[key] : undefined,
+    ) as [
+      DenormalizeNullable<E['schema']>,
+      boolean,
+      boolean,
+      Record<string, Record<string, any>>,
+    ];
+
+    let expiresAt = selectMeta(state, key)?.expiresAt;
+    // fallback to entity expiry time
+    if (!expiresAt) {
+      // expiresAt existance is equivalent to cacheResults
+      if (found) {
+        // oldest entity dictates age
+        expiresAt = Infinity;
+        // using Object.keys ensures we don't hit `toString` type members
+        Object.keys(resolvedEntities).forEach(key =>
+          Object.keys(resolvedEntities[key]).forEach(pk => {
```

Comment:
```suggestion
        Object.entries(resolvedEntities).forEach(([key, entities]) =>
          entities.forEach(pk => {
```

Reply:
object doesn't seem to have a builtin foreach...

Reply:
oh duh, my bad on that line then